torch/_inductor/codegen/cpp_bmm_template.py

```diff
@@ -0,0 +1,285 @@
+            {%- if template.should_pack_weights %}
+            &{{kernel.index(BW, ["b_start", 0, 0, 0])}},
+            {%- else %}
+            &{{kernel.index(BW, ["b_start", 0, 0])}},
+            {%- endif %}
+            &{{kernel.index(BY, ["b_start", 0, 0])}}
+            {%- if is_dynamic_M %},
+            {{kernel.size(GemmOut, -2)}}
+            {%- endif %}
+        );
+    }
+}
+"""
+
+
+class CppBmmTemplate(CppPackedGemmTemplate):
+    def __init__(
+        self,
+        input_nodes,
+        layout: ir.Layout,
+        num_threads: int,
+        register_blocking: GemmBlocking,
+        beta=1,
+        alpha=1,
+        has_bias=False,
+        epilogue_creator: Optional[Callable[[ir.Buffer], ir.Pointwise]] = None,
+        name="bmm",
+    ):
+        super().__init__(
+            input_nodes,
+            layout,
+            num_threads,
+            register_blocking,
+            beta=beta,
+            alpha=alpha,
+            has_bias=has_bias,
+            epilogue_creator=epilogue_creator,
+            name=name,
+        )
+        # Value may be changed after micro_gemm is instantiated if using VNNI layout
+        self.should_pack_weights = False
+
+    @classmethod
+    def prep_weight(cls, inputs, layout_or_out, micro_gemm):
+        if isinstance(inputs[1], ir.IRNode):
+            n = inputs[1].get_size()[-1]
+        else:
+            n = inputs[1].shape[-1]
+        _, block_n, _ = micro_gemm.register_blocking
+        padded_n = get_padded_n(n, block_n)
+        if n != padded_n and micro_gemm.get_b_layout() == LayoutType.NORMAL:
+            cls._pad_weight(inputs, padding=padded_n - n)
+        elif micro_gemm.get_b_layout() != LayoutType.NORMAL:
+            inputs, layout_or_out = cls._pack_weight(inputs, layout_or_out, micro_gemm)
+        return inputs, layout_or_out
+
+    @staticmethod
+    def _pad_weight(inputs, padding):
+        W = inputs[1]
+        if isinstance(W, ir.IRNode):
+            padded_w = L.constant_pad_nd(W, (0, padding))
+            padded_w = ir.ExternKernel.realize_input(padded_w)
+            padded_w = ir.ExternKernel.require_contiguous(padded_w)
+            if isinstance(padded_w, ir.ReinterpretView):
+                # normalize stride to be "contiguous_strides" per size
+                # this avoids the problems in L.view during template codegen
+                assert isinstance(padded_w.layout, ir.FixedLayout)
+                padded_w.layout = ir.FixedLayout(
+                    padded_w.layout.device,
+                    padded_w.layout.dtype,
+                    padded_w.layout.size,
+                    ir.FlexibleLayout.contiguous_strides(padded_w.layout.size),
+                    padded_w.layout.offset,
+                )
+        else:
+            padded_w = torch.nn.functional.pad(W, (0, padding))
+        inputs[1] = padded_w
+
+    @staticmethod
+    def _pack_weight(inputs, layout_or_out, micro_gemm):
```

Comment:
Seems a lot of duplication with cpp_gemm_template. Can we refactor/share code?

Reply:
I did a significant refactor with pack_weight and blocking for VNNI and IRNodes separately.